libs/src/services/creatorNode/CreatorNodeSelection.js

```diff
@@ -0,0 +1,249 @@
+    this.decisionTree.push({ stage: 'Made A Selection of a Primary and Secondaries', val: { primary, secondaries: secondaries.toString() } })
+
+    console.info('CreatorNodeSelection - final decision tree state', this.decisionTree)
+
+    return { primary, secondaries, services: healthCheckedServices }
+  }
+
+  /**
+   * Since all candidates are added as a backup, remove services that cleared the health
+   * check and sync status check from backups. This will prevent duplicate services from
+   * getting chosen as a backup.
+   * @param {[string]} services
+   */
+  async removePotentialPrimariesFromBackups (services) {
+    Promise.all(services.map(service => {
+      this.removeFromBackups(service)
+      return Promise.resolve()
+    }))
+  }
+
+  /**
+   * Fetches multiple urls and times each request and returns the results sorted by
+   * lowest-latency.
+   * @param {Array<Object>} requests [{id, url}, {id, url}]
+   * @returns { Array<{url, response, millis}> }
+   */
+  async timeRequests (requests) {
+    let timings = await Promise.all(requests.map(async request =>
+      timeRequest(request)
+    ))
+
+    this.sortBySemver(timings)
+    return timings
+  }
+
+  /**
+   * Sort by highest version number. If the compared services are of the same version, sort by
+   * its response time.
+   * @param {{ request ({url: 'get_request_url'}), response (axios response), millis (ms)}} services
+   */
+  sortBySemver (services) {
+    services.sort((a, b) => {
+      try {
+        if (semver.gt(a.response.data.data.version, b.response.data.data.version)) return -1
+        if (semver.lt(a.response.data.data.version, b.response.data.data.version)) return 1
+      } catch (e) {
+        // Unable to sort by version -- probably failed health check. Send to the back
+        if (!a.response) return 1
+        if (!b.response) return -1
+      }
+
+      // If same version, do a tie breaker on the response time
+      return a.millis - b.millis
+    })
+  }
+
+  getPrimary (services) {
+    // Index 0 of services will be the most optimal creator node candidate
+    // Select 1 from healthy services
+    let primary = services[0]
+    // All nodes are stored as backup unless unhealthy. Once chosen, remove from backups
+
+    // If no healthy services available, select from backups
+    if (!primary) {
+      primary = this.selectFromBackups()
+    }
+
+    // If no backups available, select from unhealthy
+    if (!primary) {
+      primary = this.selectFromUnhealthy()
+    }
+
+    return primary
+  }
+
+  getSecondaries (services, primary) {
+    // Index 1 to n of services will be sorted in highest version -> lowest version
+    // Select up to numberOfNodes-1 of secondaries that is not null and not the primary
+    let remainingSecondaries = this.numberOfNodes - 1
+    const secondaries = services
+      .filter(service => {
+        const validSecondary = service && service !== primary && remainingSecondaries-- > 0
```

Comment:
remainingSecondaries-- > 0 not inside a loop is kinda weird -- should it just be remainingSecondaries > 0?

Reply:
:3 filter is a loop